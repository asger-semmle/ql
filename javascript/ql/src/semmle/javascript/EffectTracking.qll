/**
 * Provides classes and predicates for reasoning about the ordering of side effects
 * on a value.
 */
import javascript

module EffectTracking {
  private import semmle.javascript.dataflow.internal.FlowSteps as FlowSteps

  abstract class Configuration extends string {
    bindingset[this]
    Configuration() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * in a way that matters for this effect tracking configuration.
     */
    abstract predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value);

    /**
     * Holds if a function call itself should be marked as a side effect if the
     * callee has side effects.
     */
    predicate shouldSummarizeCalls() { any() }

    /**
     * Holds if the operation at the given location has an effect on the given value,
     * or (transitively) calls a function that performs side effects on the value.
     */
    predicate hasSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      isTrackedSideEffect(loc, value) or
      calleeHasSideEffects(DataFlow::valueNode(loc), value)
    }

    /**
     * Holds if the given function has a side effect that manipulates `value`.
     */
    predicate functionHasSideEffects(Function f, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc |
        hasSideEffect(loc, value) and
        loc.getContainer() = f)
    }

    /**
     * Holds if the function invoked by `invoke` performs side effects on the given value
     * passed into the call.
     */
    predicate calleeHasSideEffects(DataFlow::InvokeNode invoke, DataFlow::SourceNode value) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (DataFlow::Node arg, Function target, Parameter param |
          FlowSteps::argumentPassing(invoke, arg, target, param) and
          value.flowsTo(arg) and
          functionHasSideEffects(target, DataFlow::parameterNode(param)))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          functionHasSideEffects(target, value))
      )
    }

    /**
     * Holds if the given basic block has a effect on the given value.
     */
    private predicate basicBlockHasSideEffects(ReachableBasicBlock block, DataFlow::SourceNode value) {
      exists (ControlFlowNode loc | 
        loc.getBasicBlock() = block and
        hasSideEffect(loc, value))
    }

    /**
     * Gets the last side effect to reach `loc`, unless `loc` is an internal node in a basic block
     * without any side effects.
     *
     * In the latter case, the last effect should be obtained using `getSummarizedSideEffectInBasicBlock`.
     */
    private ControlFlowNode getLastSideEffectAtInternalNode(ControlFlowNode loc, DataFlow::SourceNode value) {
      if hasSideEffect(loc, value) then
        result = loc
      else if isStartOfBasicBlock(loc) then
        exists (ControlFlowNode pred | pred = loc.getAPredecessor() |
          if basicBlockHasSideEffects(pred.getBasicBlock(), value) then
            result = getLastSideEffectAtInternalNode(pred, value)
          else
            result = getLastSideEffectAtInternalNode(pred.getBasicBlock().getFirstNode(), value))
      else if basicBlockHasSideEffects(loc.getBasicBlock(), value) then
        result = getLastSideEffectAtInternalNode(loc.getAPredecessor(), value)
      else
        none() // don't store the recent effect on internal nodes
    }

    /**
     * Gets the last side effect to reach the given basic block, assuming the block has no side effects of its own.
     */
    private ControlFlowNode getSummarizedSideEffectInBasicBlock(BasicBlock block, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(block.getFirstNode(), value)
      and not basicBlockHasSideEffects(block, value)
    }

    /**
     * Gets the last side effect to affect the given value at the given location.
     */
    pragma[inline]
    ControlFlowNode getLastSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      result = getLastSideEffectAtInternalNode(loc, value)
      or
      result = getSummarizedSideEffectInBasicBlock(loc.getBasicBlock(), value)
    }

    /**
     * Gets the last operation to affect the given value before the given function returns.
     */
    ControlFlowNode getLastSideEffectInFunction(Function f, DataFlow::SourceNode value) {
      result = getLastSideEffect(f.getExit(), value)
    }

    /**
     * Gets the last operation to affect the given value inside the callee of `invoke`.
     *
     * `value` refers to the value passed to the call, and `calleeAlias` refers to its local alias in the callee.
     */
    ControlFlowNode getLastSideEffectInCallee(DataFlow::InvokeNode invoke, DataFlow::SourceNode value, DataFlow::SourceNode calleeAlias) {
      shouldSummarizeCalls() and
      (
        // Passed as argument
        exists (Function target, Parameter parm, DataFlow::Node arg |
          FlowSteps::argumentPassing(invoke, arg, target, parm) and
          value.flowsTo(arg) and
          calleeAlias = DataFlow::parameterNode(parm) and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Returned from function
        exists (Function target |
          target = invoke.getACallee() and
          value = invoke and
          calleeAlias = target.getAReturnedExpr().flow().getALocalSource() and
          result = getLastSideEffectInFunction(target, calleeAlias))
        or
        // Captured by nested function
        exists (Function target |
          target = invoke.getACallee() and
          value.getContainer() != target and
          result = getLastSideEffectInFunction(target, value) and
          calleeAlias = value)
      )
    }
  }

  private predicate isStartOfBasicBlock(ControlFlowNode node) {
    exists (BasicBlock b | b.getFirstNode() = node)
  }
}
