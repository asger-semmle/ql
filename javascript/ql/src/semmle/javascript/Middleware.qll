/**
 * Provides generic concepts for modelling the middleware stack of a web application.
 */
import javascript
import semmle.javascript.EffectTracking
import semmle.javascript.frameworks.ExpressMiddleware // imported from here to avoid empty recursion error

module Middleware {
  /**
   * An effect-tracking configuration, tracking installations of route handlers on a router.
   */
  class SetupTracking extends EffectTracking::Configuration {
    SetupTracking() { this = "MiddlewareSetup" }
  
    override predicate isTrackedSideEffect(ControlFlowNode loc, DataFlow::SourceNode value) {
      exists (Setup setup |
        loc = setup.getControlFlowNode() and
        value = setup.getRouter())
    }
  }

  /**
   * A node in a web application's middleware stack model.
   *
   * A node very loosely represents a thing that can handle requests.
   * Incoming requests come from the parent and will be routed to the children,
   * in order of first to last.
   */
  abstract class Node extends DataFlow::Node {
    final ControlFlowNode getControlFlowNode() {
      result = getAstNode()
    }

    /**
     * Gets the local source of the router whose middleware stack is modeled by this node.
     *
     * For routers, this refers to the node itself.
     *
     * For setups that imperatively install route handlers on an existing router, this
     * refers to the router that is being modified.
     */
    abstract DataFlow::SourceNode getRouter();

    /**
     * Gets the last child of this node.
     */
    abstract Node getLastChild();

    /**
     * Gets the previous sibling of this node, that is, the router handler executing
     * before this one.
     *
     * In general there can be multiple previous siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    abstract Node getPreviousSibling();

    /**
     * Gets the next sibling of this node, that is, the router handler executing
     * after this one.
     *
     * In general there can be multiple next siblings, forming a sibling graph,
     * but in practice there is rarely more than once.
     */
    final Node getNextSibling() {
      this = result.getPreviousSibling()
    }

    /**
     * Gets a child of this node, that is, a router handler that is executed
     * as part of this route handler.
     */
    final Node getAChild() {
      result = this.getLastChild().getPreviousSibling*()
    }
    
    /**
     * Gets the parent of this node, that is, the route handler that dispatches
     * requests to this one.
     */
    final Node getParent() {
      this = result.getAChild()
    }

    /**
     * Gets the path matched by this route handler, relative to its parent.
     *
     * A route handler will reject a request not matching its path.
     *
     * Defaults to the empty string, meaning any request is accepted.
     */
    string getPath() {
      result = ""
    }

    /**
     * Normalizes a path so multiple paths can be joined by concatenation.
     *
     * In particular, this ensures we don't get duplicate `//` path separators.
     */
    bindingset[str]
    private string normalizePath(string str) {
      if str = "/" or str = "" then
        result = ""
      else if str = any(string s) + "/" then
        result = str
      else
        result = str + "/"
    }

    /**
     * Gets the path handled by this node, relative to its parent.
     */
    private string getRelativePath() {
      result = normalizePath(this.getPath())
      or
      not exists(this.getPath()) and
      result = ""
    }

    /**
     * Gets the path handled by this node, relative to the root node.
     */
    final string getAbsolutePath() {
      isTree() and
      (
        result = getParent().getAbsolutePath() + getRelativePath()
        or
        not exists(getParent()) and
        result = getRelativePath()
      )
      or
      // If there are multiple parents, use our best approximation: the absolute paths of any of the parents
      not isTree() and
      result = getParent().getAbsolutePath()
    }

    /**
     * Gets the request method matched by this route handler, or `*` if all
     * request methods are matched.
     *
     * The route handler will reject requests whose request method does not match.
     *
     * Defaults to `*`.
     */
    string getRequestMethod() {
      result = "*"
    }

    /**
     * Gets the request method matched by this router handler and its parents.
     */
    final string getAbsoluteRequestMethod() {
      isRoot() and result = getRequestMethod()
      or
      result = intersectRequestMethodFilters(getParent().getRequestMethod(), getRequestMethod())
    }

    /**
     * Holds if this has no parent.
     */
    final predicate isRoot() {
      not exists(getParent())
    }

    /**
     * Holds if this is part of a tree-shaped middleware stack, which is the common case.
     */
    final predicate isTree() {
      isRoot()
      or
      strictcount(getParent()) = 1 and
      getParent().isTree()
    }

    /**
     * Holds if this handler is preceded by `guard` in the middleware stack.
     *
     * That is, any request seen by this handler is known to have gone through `guard` first.
     */
    final predicate isGuardedBy(Node guard) {
      getParent*().getPreviousSibling+().getAChild*() = guard and
      getAbsolutePath() = guard.getAbsolutePath() + any(string s) and
      exists(intersectRequestMethodFilters(getAbsoluteRequestMethod(), guard.getAbsoluteRequestMethod()))
    }
  }

  /**
   * Gets the intersection of two request method filters.
   */
  bindingset[method1, method2]
  private string intersectRequestMethodFilters(string method1, string method2) {
    method1 = method2 and result = method1
    or
    method1 = "*" and result = method2
    or
    method2 = "*" and result = method1
  }

  /**
   * A middleware node with an indexed list of children.
   *
   * For example, a route setup `app.get("/", handler1, handler2, ...)` has a list of
   * children derived from its argument list.
   */
  private abstract class SequenceNode extends Node {
    /**
     * Gets the `n`th middleware function installed at this point.
     */
    abstract DataFlow::Node getMiddleware(int n);

    /**
     * Gets any of the middleware functions installed at this point.
     */
    final DataFlow::Node getAMiddleware() { result = getMiddleware(_) }

    final int getNumMiddleware() {
      result = count(getAMiddleware())
    }

    final DataFlow::Node getLastMiddleware() {
      result = getMiddleware(getNumMiddleware() - 1)
    }

    override final Node getLastChild() {
      result = getMiddleware(getNumMiddleware() - 1)
    }
  }

  /**
   * An argument in a sequence node.
   */
  private class SequenceElement extends Node {
    SequenceNode parent;
    int index;

    SequenceElement() {
      this = parent.getMiddleware(index)
    }

    override DataFlow::SourceNode getRouter() {
      result = parent.getRouter()
    }
    
    override Node getPreviousSibling() {
      result = parent.getMiddleware(index - 1)
    }
    
    override Node getLastChild() {
      result != this and
      (
        result.(DataFlow::SourceNode).flowsTo(this)
        or
        result.(DataFlow::TrackedNode).flowsTo(this)
      )
    }
  }

  /**
   * An installation of one or more route handlers on a router.
   *
   * The router handler arguments are the children of this node.
   */
  abstract class Setup extends SequenceNode {
    override final Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), getRouter())
    }
  }

  /**
   * A router, that is, a collection of route handlers with associated paths and request methods.
   */
  class Router extends Node {
    Router() {
      this = any(Node node).getRouter()
    }

    override DataFlow::SourceNode getRouter() {
      result = this
    }
    
    override Node getPreviousSibling() {
      none()
    }
    
    override Node getLastChild() {
      result.(Setup).getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(this.getContainer().getExit(), this)
    }
  }

  /**
   * A functionally defined router, as opposed to routers built up imperatively by adding router handlers on them.
   *
   * For example, an array of route handlers can in some frameworks be used as a route handler itself.
   * In this case, an array literal `[r1, r2]` would be considered a route combinator, with `r1` and `r2` as
   * its children.
   *
   * A route combinator is a `Setup` with themselves as the router.
   */
  abstract class RouteCombinator extends SequenceNode {
    override DataFlow::SourceNode getRouter() {
      result = this
    }

    override Middleware::Node getPreviousSibling() {
      none()
    }
  }

  /**
   * An array literal passed to a route setup.
   */
  private class ArrayRouteCombinator extends Middleware::RouteCombinator, DataFlow::ArrayCreationNode {
    ArrayRouteCombinator() {
      this.flowsTo(any(Node node))
    }
  
    override DataFlow::Node getMiddleware(int n) {
      result = getElement(n)
    }
  }

  /**
   * A call to a function that contains middleware installations.
   */
  class SubSetup extends Middleware::Node, DataFlow::InvokeNode {
    DataFlow::SourceNode router;

    SubSetup() {
      any(SetupTracking tr).calleeHasSideEffects(this, router)
    }

    override DataFlow::SourceNode getRouter() {
      result = router
    }

    override Middleware::Node getLastChild() {
      exists (DataFlow::SourceNode subrouter, ControlFlowNode loc |
        loc = any(SetupTracking tr).getLastSideEffectInCallee(this, router, subrouter) and
        if subrouter = router then
          result.getControlFlowNode() = loc
        else
          result = subrouter)
    }

    override Middleware::Node getPreviousSibling() {
      result.getControlFlowNode() = any(SetupTracking tr).getLastSideEffect(getControlFlowNode().getAPredecessor(), router)
    }
  }
}
