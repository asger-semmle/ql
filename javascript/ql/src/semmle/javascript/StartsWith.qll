/**
 * Provides classes and predicates for recognizing `startsWith` calls and their equivalents.
 */
import javascript

/**
 * A expression that is equivalent to `A.startsWith(B)` or `!A.startsWith(B)`.
 */
abstract class StartsWithCheck extends DataFlow::Node {
  /**
   * Gets the `A` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getBaseString();

  /**
   * Gets the `B` in `A.startsWith(B)`.
   */
  abstract DataFlow::Node getSubstring();

  /**
   * Gets the polarity if the check.
   *
   * If the polarity is `false` the check returns `true` if the string does not start
   * with the given substring.
   */
  boolean getPolarity() { result = true }
}

/**
 * An expression of form `A.startsWith(B)`.
 */
private class NativeStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  NativeStartsWith() {
    getMethodName() = "startsWith" and
    getNumArgument() = 1
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }
}

/**
 * An expression of form `A.indexOf(B) === 0`.
 */
private class IndexOfStartsWith extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::MethodCallNode indexOf;

  IndexOfStartsWith() {
    indexOf.getMethodName() = "indexOf" and
    indexOf.getNumArgument() = 1 and
    indexOf.flowsToExpr(astNode.getAnOperand()) and
    astNode.getAnOperand().getIntValue() = 0
  }

  override DataFlow::Node getBaseString() {
    result = indexOf.getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = indexOf.getArgument(0)
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}

/**
 * An expression of form `A.indexOf(B)` coerced to a boolean.
 *
 * This is equivalent to `!A.startsWith(B)` since all return values other than zero map to `true`.
 */
private class IndexOfCoercionStartsWith extends StartsWithCheck, DataFlow::MethodCallNode {
  IndexOfCoercionStartsWith() {
    getMethodName() = "indexOf" and
    getNumArgument() = 1 and
    this.flowsToExpr(any(ConditionGuardNode guard).getTest()) // check for boolean coercion
  }

  override DataFlow::Node getBaseString() {
    result = getReceiver()
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(0)
  }

  override boolean getPolarity() {
    result = false
  }
}

/**
 * A call of form `_.startsWith(A, B)` or `ramda.startsWith(A, B)`.
 */
private class LibraryStartsWith extends StartsWithCheck, DataFlow::CallNode {
  LibraryStartsWith() {
    getNumArgument() = 2 and
    exists (DataFlow::SourceNode callee | this = callee.getACall() |
      callee = LodashUnderscore::member("startsWith") or
      callee = DataFlow::moduleMember("ramda", "startsWith")
    )
  }

  override DataFlow::Node getBaseString() {
    result = getArgument(0)
  }

  override DataFlow::Node getSubstring() {
    result = getArgument(1)
  }
}

/**
 * A comparison of form `x[0] === "k"` for some single-character constant `k`.
 */
private class FirstCharacterCheck extends StartsWithCheck, DataFlow::ValueNode {
  override EqualityTest astNode;
  DataFlow::PropRead read;
  Expr constant;
  
  FirstCharacterCheck() {
    read.flowsTo(astNode.getAnOperand().flow()) and
    read.getPropertyNameExpr().getIntValue() = 0 and
    constant.getStringValue().length() = 1 and
    astNode.getAnOperand() = constant
  }

  override DataFlow::Node getBaseString() {
    result = read.getBase()
  }

  override DataFlow::Node getSubstring() {
    result = constant.flow()
  }

  override boolean getPolarity() {
    result = astNode.getPolarity()
  }
}
