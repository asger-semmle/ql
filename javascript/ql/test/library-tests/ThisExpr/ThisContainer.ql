import javascript

from DataFlow::ThisNode node
select node, node.getBasicBlock(), node.getContainer()
