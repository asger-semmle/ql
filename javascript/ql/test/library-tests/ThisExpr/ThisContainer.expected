| tst.js:1:16:1:15 | this | tst.js:1:16:1:15 | entry node of () {} | tst.js:1:16:1:15 | () {} |
| tst.js:3:7:3:6 | this | tst.js:3:7:3:6 | entry node of () {\\n   ... ;\\n    } | tst.js:3:7:5:5 | () {\\n   ... ;\\n    } |
| tst.js:7:7:7:6 | this | tst.js:7:7:7:6 | entry node of () {\\n   ... ;\\n    } | tst.js:7:7:9:5 | () {\\n   ... ;\\n    } |
| tst.js:13:14:13:13 | this | tst.js:13:14:13:13 | entry node of () {} | tst.js:13:14:13:13 | () {} |
| tst.js:15:13:15:12 | this | tst.js:15:13:15:12 | entry node of () {\\n   ... ;\\n    } | tst.js:15:13:19:5 | () {\\n   ... ;\\n    } |
| tst.js:16:10:16:9 | this | tst.js:16:10:16:9 | entry node of functio ...       } | tst.js:16:10:18:9 | functio ...       } |
| tst.js:21:13:21:12 | this | tst.js:21:13:21:12 | entry node of () {\\n   ... ;\\n    } | tst.js:21:13:26:5 | () {\\n   ... ;\\n    } |
| tst.js:23:10:23:9 | this | tst.js:23:10:23:9 | entry node of functio ...       } | tst.js:23:10:25:9 | functio ...       } |
| tst.js:28:13:28:12 | this | tst.js:28:13:28:12 | entry node of () {\\n   ... ;\\n    } | tst.js:28:13:33:5 | () {\\n   ... ;\\n    } |
| tst.js:29:9:29:8 | this | tst.js:29:9:29:8 | entry node of functio ...       } | tst.js:29:9:31:9 | functio ...       } |
| tst.js:35:13:35:12 | this | tst.js:35:13:35:12 | entry node of () {\\n   ... ;\\n    } | tst.js:35:13:41:5 | () {\\n   ... ;\\n    } |
| tst.js:36:10:36:9 | this | tst.js:36:10:36:9 | entry node of functio ...       } | tst.js:36:10:40:9 | functio ...       } |
| tst.js:37:14:37:13 | this | tst.js:37:14:37:13 | entry node of functio ...       } | tst.js:37:14:39:13 | functio ...       } |
| tst.js:45:14:45:13 | this | tst.js:45:14:45:13 | entry node of () {} | tst.js:45:14:45:13 | () {} |
| tst.js:47:13:47:12 | this | tst.js:47:13:47:12 | entry node of () {\\n   ... ;\\n    } | tst.js:47:13:51:5 | () {\\n   ... ;\\n    } |
| tst.js:48:10:48:9 | this | tst.js:48:10:48:9 | entry node of functio ...       } | tst.js:48:10:50:9 | functio ...       } |
| tst.js:55:15:55:14 | this | tst.js:55:15:55:14 | entry node of () {} | tst.js:55:15:55:14 | () {} |
| tst.js:57:13:57:12 | this | tst.js:57:13:57:12 | entry node of () {\\n   ... ;\\n    } | tst.js:57:13:62:5 | () {\\n   ... ;\\n    } |
| tst.js:59:10:59:9 | this | tst.js:59:10:59:9 | entry node of functio ...       } | tst.js:59:10:61:9 | functio ...       } |
| tst.js:66:13:66:12 | this | tst.js:66:13:66:12 | entry node of () {} | tst.js:66:13:66:12 | () {} |
| tst.js:68:13:68:12 | this | tst.js:68:13:68:12 | entry node of () {\\n   ... ;\\n    } | tst.js:68:13:72:5 | () {\\n   ... ;\\n    } |
| tst.js:69:15:69:14 | this | tst.js:69:15:69:14 | entry node of functio ...       } | tst.js:69:15:71:9 | functio ...       } |
| tst.js:74:13:74:12 | this | tst.js:74:13:74:12 | entry node of () {\\n   ... )\\n    } | tst.js:74:13:79:5 | () {\\n   ... )\\n    } |
| tst.js:76:10:76:9 | this | tst.js:76:10:76:9 | entry node of functio ...       } | tst.js:76:10:78:9 | functio ...       } |
| tst.js:83:22:83:21 | this | tst.js:83:22:83:21 | entry node of () {} | tst.js:83:22:83:21 | () {} |
| tst.js:85:13:85:12 | this | tst.js:85:13:85:12 | entry node of () {\\n   ... `\\n    } | tst.js:85:13:93:5 | () {\\n   ... `\\n    } |
| tst.js:87:21:87:20 | this | tst.js:87:21:87:20 | entry node of functio ...       } | tst.js:87:21:89:13 | functio ...       } |
| tst.js:95:13:95:12 | this | tst.js:95:13:95:12 | entry node of () {\\n   ... ;\\n    } | tst.js:95:13:100:5 | () {\\n   ... ;\\n    } |
| tst.js:96:9:96:8 | this | tst.js:96:9:96:8 | entry node of functio ...       } | tst.js:96:9:98:9 | functio ...       } |
| tst.js:105:20:105:19 | this | tst.js:105:20:105:19 | entry node of () {} | tst.js:105:20:105:19 | () {} |
| tst.js:107:13:107:12 | this | tst.js:107:13:107:12 | entry node of () {\\n   ... ;\\n    } | tst.js:107:13:111:5 | () {\\n   ... ;\\n    } |
| tst.js:108:18:108:17 | this | tst.js:108:18:108:17 | entry node of functio ...       } | tst.js:108:18:110:9 | functio ...       } |
| tst.js:113:13:113:12 | this | tst.js:113:13:113:12 | entry node of () {\\n   ... ;\\n    } | tst.js:113:13:117:5 | () {\\n   ... ;\\n    } |
| tst.js:114:21:114:20 | this | tst.js:114:21:114:20 | entry node of functio ...       } | tst.js:114:21:116:9 | functio ...       } |
| tst.js:119:13:119:12 | this | tst.js:119:13:119:12 | entry node of () {\\n   ... ;\\n    } | tst.js:119:13:123:5 | () {\\n   ... ;\\n    } |
| tst.js:120:27:120:26 | this | tst.js:120:27:120:26 | entry node of functio ...       } | tst.js:120:27:122:9 | functio ...       } |
| tst.js:128:16:128:15 | this | tst.js:128:16:128:15 | entry node of () {} | tst.js:128:16:128:15 | () {} |
| tst.js:130:13:130:12 | this | tst.js:130:13:130:12 | entry node of () {\\n   ... ;\\n    } | tst.js:130:13:134:5 | () {\\n   ... ;\\n    } |
| tst.js:131:18:131:17 | this | tst.js:131:18:131:17 | entry node of functio ...       } | tst.js:131:18:133:9 | functio ...       } |
| tst.js:136:13:136:12 | this | tst.js:136:13:136:12 | entry node of () {\\n   ... ;\\n    } | tst.js:136:13:140:5 | () {\\n   ... ;\\n    } |
| tst.js:137:21:137:20 | this | tst.js:137:21:137:20 | entry node of functio ...       } | tst.js:137:21:139:9 | functio ...       } |
| tst.js:142:13:142:12 | this | tst.js:142:13:142:12 | entry node of () {\\n   ... ;\\n    } | tst.js:142:13:146:5 | () {\\n   ... ;\\n    } |
| tst.js:143:23:143:22 | this | tst.js:143:23:143:22 | entry node of functio ...       } | tst.js:143:23:145:9 | functio ...       } |
